src/main/java/org/embulk/input/postgresql_wal/PostgresqlWalDumper.java

```diff
@@ -0,0 +1,141 @@
+import org.embulk.input.postgresql_wal.decoders.Wal2JsonDecoderPlugin;
+import org.embulk.input.postgresql_wal.model.*;
+import org.embulk.spi.PageBuilder;
+import org.embulk.spi.Schema;
+import org.postgresql.replication.LogSequenceNumber;
+import org.postgresql.replication.PGReplicationStream;
+
+import java.nio.ByteBuffer;
+import java.sql.Connection;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+
+public class PostgresqlWalDumper {
+    private PluginTask task;
+    private PageBuilder pageBuilder;
+    private Schema schema;
+    private ConnectionManager connectionManager;
+    private PostgresqlWalClient walClient;
+    private Connection connection;
+    private Wal2JsonDecoderPlugin decoderPlugin;
+    private LogSequenceNumber lsn;
+
+    public PostgresqlWalDumper(PluginTask task, PageBuilder pageBuilder, Schema schema, Connection connection) {
+        this.task = task;
+        this.pageBuilder = pageBuilder;
+        this.schema = schema;
+        this.connection = connection;
+        this.decoderPlugin = new Wal2JsonDecoderPlugin();
+    }
+
+    public void start() {
+        try {
+            walClient = new PostgresqlWalClient(connection);
+
+            PGReplicationStream stream = walClient.getReplicationStream(task.getSlot());
+            long waitMin = task.getWalInitialWait();
+            int retryCount = 0;
+            long wait = waitMin;
+            while (!stream.isClosed()) {
+                ByteBuffer msg = stream.readPending(); // non-blocking
+                // stop if exceed LSN or time out
+                if (msg == null) {
+                    TimeUnit.MILLISECONDS.sleep(wait);
+                    retryCount += 1;
+                    wait = waitMin * (long)Math.pow(2, retryCount);
+                    if (wait < task.getWalReadTimeout()){
+                        continue;
+                    } else {
+                        break;
+                    }
+                }else{
+                    retryCount = 0;
+                    wait = waitMin;
+                }
+
+                List<AbstractRowEvent> rowEvents = decoderPlugin.decode(msg, stream.getLastReceiveLSN());
+                for (AbstractRowEvent rowEvent: rowEvents) {
+                    LsnHolder.setLsn(rowEvent.getNextLogSequenceNumber());
+                    handleRowEvent(rowEvent);
+                }
+
+                if (task.getToLsn().isPresent()){
+                    if (LsnHolder.getLsn().asLong() >= LogSequenceNumber.valueOf(task.getToLsn().get()).asLong()){
+                        break;
+                    }
+                }
+            }
+            if (!stream.isClosed()){
+                stream.close();
+            }
+            if (!ConnectionManager.getReplicationConnection().isClosed()){
+                ConnectionManager.getReplicationConnection().close();
+            }
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
```

Comment:
need to `stream.close` with `finally` block? what do you think?

Reply:
a8d6e64

DONE